File: clients/ios/Lextures/Core/LMS/LiveQuizLogic.swift
import Foundation

/// Pure helpers for interactive live quizzes (MOB.5 Phase 1 — student play).
enum LiveQuizLogic {
    static let nicknameMaxLength = 24
    private static let nicknameAllowed = try! NSRegularExpression(
        pattern: #"^[\p{L}\p{N} _.\-'!]+$"#
    )

    enum JoinStep: String, Equatable {
        case code
        case nickname
        case play
    }

    enum NicknameReason: String, Equatable {
        case empty
        case tooLong
        case charset
    }

    enum NicknameValidation: Equatable {
        case ok(String)
        case invalid(NicknameReason)
    }

    enum JoinErrorReason: String, Equatable {
        case notFound
        case rateLimited
        case nicknameTaken
        case nicknameInvalid
        case nicknameDenied
        case lobbyLocked
        case banned
        case oneSession
        case gameEnded
        case authRequired
        case joinFailed
        case unknown
    }

    static func normalizeJoinCode(_ raw: String) -> String {
        raw.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
    }

    static func isValidJoinCode(_ raw: String) -> Bool {
        let code = normalizeJoinCode(raw)
        guard !code.isEmpty, code.count <= 12 else { return false }
        return code.allSatisfy { $0.isLetter || $0.isNumber }
    }

    static func normalizeNickname(_ raw: String) -> String {
        raw.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func validateNickname(_ raw: String) -> NicknameValidation {
        let nickname = normalizeNickname(raw)
        if nickname.isEmpty { return .invalid(.empty) }
        if nickname.count > nicknameMaxLength { return .invalid(.tooLong) }
        let range = NSRange(nickname.startIndex..<nickname.endIndex, in: nickname)
        guard nicknameAllowed.firstMatch(in: nickname, options: [], range: range) != nil else {
            return .invalid(.charset)
        }
        return .ok(nickname)
    }

    /// Dual gate: per-course interactive quizzes + mobile rollout kill-switch.
    static func shouldShowWorkspaceSection(
        course: CourseSummary,
        features: MobilePlatformFeatures
    ) -> Bool {
        course.isInteractiveQuizzesEnabled && features.ffMobileLiveQuiz
    }

    static func liveQuizEntryEnabled(
        courseEnabled: Bool,
        features: MobilePlatformFeatures
    ) -> Bool {
        courseEnabled && features.ffMobileLiveQuiz
    }

    static func joinErrorReason(status: Int, body: String) -> JoinErrorReason {
        let lower = body.lowercased()
        switch status {
        case 404:
            return .notFound
        case 429:
            return .rateLimited
        case 409:
            if lower.contains("already connected") { return .oneSession }
            return .nicknameTaken
        case 401, 403:
            if lower.contains("lobby") { return .lobbyLocked }
            if lower.contains("rejoin") || lower.contains("cannot") { return .banned }
            return .authRequired
        case 400:
            if lower.contains("ended") { return .gameEnded }
            if lower.contains("isn’t allowed") || lower.contains("isn't allowed") || lower.contains("not allowed") {
                return .nicknameDenied
            }
            if lower.contains("nickname") { return .nicknameInvalid }
            return .joinFailed
        default:
            return .unknown
        }
    }

    static func joinErrorLocalizationKey(_ reason: JoinErrorReason) -> String {
        switch reason {
        case .notFound: return "mobile.liveQuiz.error.notFound"
        case .rateLimited: return "mobile.liveQuiz.error.rateLimited"
        case .nicknameTaken: return "mobile.liveQuiz.error.nicknameTaken"
        case .nicknameInvalid: return "mobile.liveQuiz.error.nicknameInvalid"
        case .nicknameDenied: return "mobile.liveQuiz.error.nicknameDenied"
        case .lobbyLocked: return "mobile.liveQuiz.error.lobbyLocked"
        case .banned: return "mobile.liveQuiz.error.banned"
        case .oneSession: return "mobile.liveQuiz.error.oneSession"
        case .gameEnded: return "mobile.liveQuiz.error.gameEnded"
        case .authRequired: return "mobile.liveQuiz.error.authRequired"
        case .joinFailed: return "mobile.liveQuiz.error.joinFailed"
        case .unknown: return "mobile.liveQuiz.error.generic"
        }
    }

    static func nicknameReasonLocalizationKey(_ reason: NicknameReason) -> String {
        switch reason {
        case .empty: return "mobile.liveQuiz.nickname.error.empty"
        case .tooLong: return "mobile.liveQuiz.nickname.error.tooLong"
        case .charset: return "mobile.liveQuiz.nickname.error.charset"
        }
    }

    static func webSocketPath(courseCode: String, gameId: String) -> String {
        "/api/v1/courses/\(encodePath(courseCode))/live-quizzes/games/\(encodePath(gameId))/ws"
    }

    private static func encodePath(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? value
    }
}

/// Player session persisted for reconnect/resume (mirrors web `live-quiz-player-storage`).
struct LiveQuizPlayerSession: Codable, Equatable {
    var gameId: String
    var courseCode: String
    var playerId: String
    var playerToken: String
    var nickname: String
    var joinCode: String
}

enum LiveQuizPlayerSessionStore {
    private static let prefix = "liveQuiz.playerSession."

    static func save(_ session: LiveQuizPlayerSession) {
        guard let data = try? JSONEncoder().encode(session) else { return }
        UserDefaults.standard.set(data, forKey: prefix + session.gameId)
    }

    static func load(gameId: String) -> LiveQuizPlayerSession? {
        guard let data = UserDefaults.standard.data(forKey: prefix + gameId) else { return nil }
        return try? JSONDecoder().decode(LiveQuizPlayerSession.self, from: data)
    }

    static func clear(gameId: String) {
        UserDefaults.standard.removeObject(forKey: prefix + gameId)
    }
}

enum LiveQuizObservability {
    private static var counters: [String: Int] = [:]
    private static let lock = NSLock()

    static func record(_ event: String, attributes: [String: String] = [:]) {
        lock.lock()
        defer { lock.unlock() }
        let key = attributes.isEmpty
            ? event
            : event + "|" + attributes.keys.sorted().map { "\($0)=\(attributes[$0] ?? "")" }.joined(separator: ",")
        counters[key, default: 0] += 1
    }

    static func count(for event: String) -> Int {
        lock.lock()
        defer { lock.unlock() }
        return counters.filter { $0.key == event || $0.key.hasPrefix(event + "|") }.values.reduce(0, +)
    }

    #if DEBUG
    static func resetForTests() {
        lock.lock()
        counters.removeAll()
        lock.unlock()
    }
    #endif
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/LiveQuizLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.navigation.MobilePlatformFeatures
import kotlinx.serialization.Serializable
import java.util.concurrent.ConcurrentHashMap

/** Pure helpers for interactive live quizzes (MOB.5 Phase 1 — student play). */
object LiveQuizLogic {
    const val NICKNAME_MAX_LENGTH = 24
    private val nicknameAllowed = Regex("""^[\p{L}\p{N} _.\-'!]+$""")

    enum class JoinStep { Code, Nickname, Play }

    enum class NicknameReason { Empty, TooLong, Charset }

    sealed class NicknameValidation {
        data class Ok(val nickname: String) : NicknameValidation()
        data class Invalid(val reason: NicknameReason) : NicknameValidation()
    }

    enum class JoinErrorReason {
        NotFound,
        RateLimited,
        NicknameTaken,
        NicknameInvalid,
        NicknameDenied,
        LobbyLocked,
        Banned,
        OneSession,
        GameEnded,
        AuthRequired,
        JoinFailed,
        Unknown,
    }

    fun normalizeJoinCode(raw: String): String =
        raw.trim().uppercase()

    fun isValidJoinCode(raw: String): Boolean {
        val code = normalizeJoinCode(raw)
        if (code.isEmpty() || code.length > 12) return false
        return code.all { it.isLetterOrDigit() }
    }

    fun normalizeNickname(raw: String): String = raw.trim()

    fun validateNickname(raw: String): NicknameValidation {
        val nickname = normalizeNickname(raw)
        if (nickname.isEmpty()) return NicknameValidation.Invalid(NicknameReason.Empty)
        if (nickname.length > NICKNAME_MAX_LENGTH) {
            return NicknameValidation.Invalid(NicknameReason.TooLong)
        }
        if (!nicknameAllowed.matches(nickname)) {
            return NicknameValidation.Invalid(NicknameReason.Charset)
        }
        return NicknameValidation.Ok(nickname)
    }

    /** Dual gate: per-course interactive quizzes + mobile rollout kill-switch. */
    fun shouldShowWorkspaceSection(
        course: CourseSummary,
        features: MobilePlatformFeatures,
    ): Boolean = course.isInteractiveQuizzesEnabled && features.ffMobileLiveQuiz

    fun liveQuizEntryEnabled(
        courseEnabled: Boolean,
        features: MobilePlatformFeatures,
    ): Boolean = courseEnabled && features.ffMobileLiveQuiz

    fun joinErrorReason(status: Int, body: String): JoinErrorReason {
        val lower = body.lowercase()
        return when (status) {
            404 -> JoinErrorReason.NotFound
            429 -> JoinErrorReason.RateLimited
            409 -> if (lower.contains("already connected")) {
                JoinErrorReason.OneSession
            } else {
                JoinErrorReason.NicknameTaken
            }
            401, 403 -> when {
                lower.contains("lobby") -> JoinErrorReason.LobbyLocked
                lower.contains("rejoin") || lower.contains("cannot") -> JoinErrorReason.Banned
                else -> JoinErrorReason.AuthRequired
            }
            400 -> when {
                lower.contains("ended") -> JoinErrorReason.GameEnded
                lower.contains("isn't allowed") || lower.contains("isn’t allowed") ||
                    lower.contains("not allowed") -> JoinErrorReason.NicknameDenied
                lower.contains("nickname") -> JoinErrorReason.NicknameInvalid
                else -> JoinErrorReason.JoinFailed
            }
            else -> JoinErrorReason.Unknown
        }
    }

    fun joinErrorLocalizationKey(reason: JoinErrorReason): String = when (reason) {
        JoinErrorReason.NotFound -> "mobile.liveQuiz.error.notFound"
        JoinErrorReason.RateLimited -> "mobile.liveQuiz.error.rateLimited"
        JoinErrorReason.NicknameTaken -> "mobile.liveQuiz.error.nicknameTaken"
        JoinErrorReason.NicknameInvalid -> "mobile.liveQuiz.error.nicknameInvalid"
        JoinErrorReason.NicknameDenied -> "mobile.liveQuiz.error.nicknameDenied"
        JoinErrorReason.LobbyLocked -> "mobile.liveQuiz.error.lobbyLocked"
        JoinErrorReason.Banned -> "mobile.liveQuiz.error.banned"
        JoinErrorReason.OneSession -> "mobile.liveQuiz.error.oneSession"
        JoinErrorReason.GameEnded -> "mobile.liveQuiz.error.gameEnded"
        JoinErrorReason.AuthRequired -> "mobile.liveQuiz.error.authRequired"
        JoinErrorReason.JoinFailed -> "mobile.liveQuiz.error.joinFailed"
        JoinErrorReason.Unknown -> "mobile.liveQuiz.error.generic"
    }

    fun nicknameReasonLocalizationKey(reason: NicknameReason): String = when (reason) {
        NicknameReason.Empty -> "mobile.liveQuiz.nickname.error.empty"
        NicknameReason.TooLong -> "mobile.liveQuiz.nickname.error.tooLong"
        NicknameReason.Charset -> "mobile.liveQuiz.nickname.error.charset"
    }

    fun webSocketPath(courseCode: String, gameId: String): String =
        "/api/v1/courses/${encodePath(courseCode)}/live-quizzes/games/${encodePath(gameId)}/ws"

    private fun encodePath(value: String): String =
        java.net.URLEncoder.encode(value, "UTF-8").replace("+", "%20")
}

@Serializable
data class LiveQuizPlayerSession(
    val gameId: String,
    val courseCode: String,
    val playerId: String,
    val playerToken: String,
    val nickname: String,
    val joinCode: String,
)

/** In-memory session store for unit tests; UI layer should persist via SharedPreferences. */
object LiveQuizPlayerSessionStore {
    private val sessions = ConcurrentHashMap<String, LiveQuizPlayerSession>()

    fun save(session: LiveQuizPlayerSession) {
        sessions[session.gameId] = session
    }

    fun load(gameId: String): LiveQuizPlayerSession? = sessions[gameId]

    fun clear(gameId: String) {
        sessions.remove(gameId)
    }

    fun resetForTests() {
        sessions.clear()
    }
}

object LiveQuizObservability {
    private val counters = ConcurrentHashMap<String, Int>()

    fun record(event: String, attributes: Map<String, String> = emptyMap()) {
        val key = if (attributes.isEmpty()) {
            event
        } else {
            event + "|" + attributes.toSortedMap().entries.joinToString(",") { "${it.key}=${it.value}" }
        }
        counters.merge(key, 1) { a, b -> a + b }
    }

    fun count(event: String): Int =
        counters.entries.filter { it.key == event || it.key.startsWith("$event|") }
            .sumOf { it.value }

    fun resetForTests() {
        counters.clear()
    }
}
